subprojects/execution/src/main/java/org/gradle/internal/execution/steps/ExecuteWorkBuildOperationFiringStep.java

```diff
@@ -39,29 +40,31 @@ public ExecuteWorkBuildOperationFiringStep(BuildOperationExecutor buildOperation
 
     @Override
     public R execute(UnitOfWork work, C context) {
-        return operation(operationContext -> {
-                R result = delegate.execute(work, context);
-                ExecuteWorkBuildOperationType.Result operationResult = new ExecuteWorkResult(result.getExecution());
-                operationContext.setResult(operationResult);
-                return result;
-            },
-            BuildOperationDescriptor
-                .displayName("Execute unit of work")
-                .details(new ExecuteWorkDetails(work)));
+        Optional<String> buildOperationWorkType = work.getBuildOperationWorkType();
+        return buildOperationWorkType.map(workType -> operation(operationContext -> {
+                    R result = delegate.execute(work, context);
+                    ExecuteWorkBuildOperationType.Result operationResult = new ExecuteWorkResult(result.getExecution());
+                    operationContext.setResult(operationResult);
+                    return result;
+                },
+                BuildOperationDescriptor
+                    .displayName("Execute unit of work")
+                    .details(new ExecuteWorkDetails(workType)))
+        ).orElseGet(() -> delegate.execute(work, context));
```

Comment:
💄 I can't help but notice that non-functional style is less readable in this case when we can do "short-circuting" with:
```
if (buildOperationWorkType.isEmpty())
    return delegate.execute(work, context);
}
```

And the linear code structure is better debuggable. I guess, I am not a fan of Optionals after all...